db: hints: add checksum to sync_point encoding

patjed41 opened the pull request:
Fixes #9405 

`sync_point` API provided with incorrect sync point id might allocate crazy amount of memory and fail with `std::bad_alloc`.

To fix this, we can check if the encoded sync point has been modified before decoding. We can achieve this by calculating a checksum before encoding, appending it to the encoded sync point, and compering it with a checksum calculated in `db::hints::decode` before decoding.

scylladb-promoter commented:
@scylladb/scylla-maint please review and trigger CI for this pull request

patjed41 commented:
@piodul, @kostja please review

benipeled commented:
CI triggered manually https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2283/

piodul commented on db/hints/sync_point.cc:
```diff
@@ -17,13 +17,15 @@
+// Format V1/V2 (encoded in base64):
+//   uint8_t 0x01/0x02 - version of format
 //   sync_point_v1 - encoded using IMR
+//   uint64_t - checksum (present only in V2)
```
Please indicate which algorithm is used for the checksum.

piodul commented on db/hints/sync_point.cc:
```diff
@@ -67,16 +69,30 @@ static std::vector<sync_point::shard_rps> decode_one_type_v1(uint16_t shard_coun
-    if (raw[0] != 1) {
+
+    if (raw[0] == 2) {
+        seastar::simple_memory_input_stream in_checksum{reinterpret_cast<const char*>(raw.end()) - sizeof(uint64_t), sizeof(uint64_t)};
```
Please introduce a constant somewhere somewhere high in the file (e.g. right after the comment with the  description of the serialization format):

```c++
static constexpr size_t checksum_size = sizeof(uint64_t);
```

and then use it in instead of `sizeof(uint64_t)`. It will be more readable that way.

piodul commented on test/rest_api/test_hinted_handoff.py:
```diff
@@ -0,0 +1,21 @@
+import urllib.parse
+import base64
+
+def test_hinted_handoff_sync_point(rest_api):
```
Please change the test name to something more descriptive. It's true that it tests the hinted handoff sync point feature, but this is not the only test case that you could come up with for it. More specifically, this test is for the checksum functionality that you introduced and it verifies that the checksum does detect invalid sync points and returns a proper error. For example, you could change it to `test_sync_point_checksum` (the fact that this is for _hinted handoff_ is implied by the file name).

piodul commented on test/rest_api/test_hinted_handoff.py:
```diff
@@ -0,0 +1,21 @@
+    resp = rest_api.send('GET', "hinted_handoff/sync_point", { "id": urllib.parse.quote(sync_point) })
+    assert resp.ok
+
+    for i in range(1, len(sync_point_b)):
```
Please add a comment which explains why we are skipping the first byte.

piodul replied:
The comment highlights that we do skip the first byte, but it doesn't answer the question why we do it.

patjed41 replied:
I've added an explanation.

piodul commented on test/rest_api/test_hinted_handoff.py:
```diff
@@ -0,0 +1,21 @@
+def test_hinted_handoff_sync_point(rest_api):
+    resp = rest_api.send('POST', "hinted_handoff/sync_point")
+    sync_point = resp.json()
+    sync_point_b = base64.b64decode(sync_point.encode('ascii'))
```
I think that you should add a comment here in order to explain why we need to partially decode the sync point.

patjed41 replied:
I think the new explanation is good enough.

piodul commented on db/hints/sync_point.cc:
```diff
@@ -67,16 +69,30 @@ static std::vector<sync_point::shard_rps> decode_one_type_v1(uint16_t shard_coun
     return ret;
 }
 
+static uint64_t get_checksum(signed char* data, size_t size) {
```
Nit: I think `calculate_checksum` would be a slightly better name. `get_checksum` could imply that maybe the checksum is encoded in the data and it "gets" it out of the encoded data.

piodul commented on db/hints/sync_point.cc:
```diff
@@ -67,16 +69,30 @@ static std::vector<sync_point::shard_rps> decode_one_type_v1(uint16_t shard_coun
+    if (raw[0] == 2) {
+        seastar::simple_memory_input_stream in_checksum{reinterpret_cast<const char*>(raw.end()) - sizeof(uint64_t), sizeof(uint64_t)};
+        uint64_t checksum = ser::serializer<uint64_t>::read(in_checksum);
+        if (checksum != get_checksum(raw.data(), raw.size() - sizeof(uint64_t))) {
```
If you apply my previous suggestion about changing the interface of `get_checksum`, you can use the [`.substr()`](https://en.cppreference.com/w/cpp/string/basic_string_view/substr) method (`sstring_view` an alias over `std::string_view`) which you can use to create a subslice of the original `string_view`:

```c++
if (checksum != calculate_checksum(raw.substr(0, raw.size() - checksum_size)) {
    // ...
}
```

patjed41 replied:
Unfortunately, it does not work, because `raw` has a type `bytes`.  Some casting is necessary here. I've made it work, but I don't think it's very clear.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2283/

patjed41 commented:
v2: made changes requested in @piodul's comments:
- changed name and interface of the `get_checksum` function
- change name of the `test_hinted_handoff_sync_point` test to `test_sync_point_checksum(`
- introduced `version_size` and `checksum_size` constants
- added some new comments and improved the old ones

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2303/

patjed41 commented:
v3: explained in comment why the first byte of encoded `sync_point` is omitted in `test_sync_point_checksum`

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2306/

patjed41 commented:
v4:
- simplified the definitions of the `sync_point::decode` and `sync_point::encode` functions by using `sstring` instead of `bytes` to seriealize and deserialize `sync_point` (this reduced number of casts and simplified some formulas)
- made `sync_point::decode` and `sync_point::encode` not dependent on the `version_size` value (these functions used the fact that `version_size == 0`)

piodul commented on db/hints/sync_point.cc:
```diff
@@ -71,7 +71,8 @@ static std::vector<sync_point::shard_rps> decode_one_type_v1(uint16_t shard_coun
 sync_point sync_point::decode(sstring_view s) {
     bytes raw = base64_decode(s);
-    seastar::simple_memory_input_stream in{reinterpret_cast<const char*>(raw.data()), raw.size()};
+    sstring_view raw_s(reinterpret_cast<char*>(raw.data()), raw.size());
```
Nitpicking over the second commit message, there is a typo: "serielize" -> "serialize".

piodul commented on db/hints/sync_point.cc:
```diff
@@ -69,13 +72,30 @@ static std::vector<sync_point::shard_rps> decode_one_type_v1(uint16_t shard_coun
-    if (version != 1) {
+    if (version == 2) {
+        if (raw_s.size() < version_size + checksum_size) {
+            throw std::runtime_error("Could not decode the sync point encoded in the V2 format - not a valid hex string");
```
It looks like you copied the old error message here which used to be thrown when `s` is empty, but I think it was slightly wrong and here we have a chance to improve it. How about:

```
Could not decode the sync point encoded in the V2 format - serialized blob is too short
```

piodul commented on db/hints/sync_point.cc:
```diff
@@ -135,13 +155,17 @@ sstring sync_point::encode() const {
+    sstring serialized{sstring::initialized_later{}, version_size + measure.size() + checksum_size};
+
+    // Encode using V2 format
+    seastar::simple_memory_output_stream out{serialized.data(), version_size + measure.size() + checksum_size};
```
How about `{serialized.data(), serialized.size()}`? The `serialized` string already has the correct length.

piodul commented on db/hints/sync_point.cc:
```diff
@@ -135,13 +155,17 @@ sstring sync_point::encode() const {
+    ser::serializer<uint64_t>::write(out, checksum);
 
-    return base64_encode(reinterpret_cast<int8_t*>(serialized.data(), serialized.size()));
+    return base64_encode(bytes(reinterpret_cast<signed char*>(serialized.data()), serialized.size()));
```
`bytes_view` instead of `bytes` should be sufficient here and will avoid a copy.

Also, why change `int8_t` to `signed char`?

patjed41 replied:
For some reason I thought that `bytes` are defined with `signed char`. Fixed.

piodul commented on db/hints/sync_point.cc:
```diff
@@ -135,12 +136,12 @@ sstring sync_point::encode() const {
     ser::serializer<sync_point_v1>::write(out, v1);
 
-    return base64_encode(serialized);
+    return base64_encode(reinterpret_cast<int8_t*>(serialized.data(), serialized.size()));
```
```
reinterpret_cast<int8_t*>(serialized.data(), serialized.size())
```

Is valid C++? `reinterpret_cast` takes only one parameter. Please make sure that all of the commits compile.

patjed41 replied:
Done.

piodul reviewed (changes requested):
Some final nitpicks from my side, but then LGTM.

piodul commented on db/hints/sync_point.cc:
```diff
@@ -133,11 +156,16 @@ sstring sync_point::encode() const {
     ser::serializer<sync_point_v1>::write(out, v1);
+    seastar::simple_memory_output_stream out_checksum{reinterpret_cast<char*>(serialized.data()), checksum_size, version_size + measure.size()};
+    uint64_t checksum = calculate_checksum(sstring_view(reinterpret_cast<char*>(serialized.data()), version_size + measure.size()));
+    ser::serializer<uint64_t>::write(out_checksum, checksum);
```
Sorry, I wrote this comment before talking about it offline and forgot to delete. Please disregard and feel free to resolve immediately.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2319/

patjed41 commented:
v5: fixed nitpicks from the latest comments

kbr-scylla commented on db/hints/sync_point.cc:
```diff
@@ -41,6 +41,8 @@ namespace hints {
 //       Flattened representation was chosen in order to save space on
```
The commit message in `db: hints: add the version_size constant` refers to some things that do not exist at the parent of this commit.

> As the V2 format of encoding sync points introduces the new checksum_size
constant, we can additionally add the version_size constant to increase
consistency and readability.

What's the "V2 format"? What is "the new checksum_size"?

patjed41 replied:
I changed this commit message. It should be clear now that it refers to the changes added in the 2 next commits.

piodul replied:
I don't agree. I think doing small refactors/cleanups first and then introducing functionality results in cleaner, more understandable diffs.

On one hand, I think the cover letter (PR description) should mention that the PR introduces the V2 format which adds a checksum, so that reviewers are surprised less when they see mentions of V2.

On the other hand, perhaps this commit message is too specific. Personally, I would avoid mentioning the `checksum_size` field, just that we want to extend the format and add one more field, so we do a cleanup of the existing field first.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2327/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2337/

avikivity commented:
Where are these things documented? and why did we not use IDL?

piodul commented:
> Where are these things documented?

It's not (yet), we have an open issue for that: https://github.com/scylladb/scylla-docs/issues/3787 .
Some context should be available here: https://github.com/scylladb/scylladb/pull/8982

> and why did we not use IDL?

The sync point description is serialized using IDL (not sure why I wrote "IMR" in the comment); the version field and the checksum is not. Not sure it would make sense to use IDL for the checksum as we need to parse it first before the rest.

avikivity commented:
> > Where are these things documented?
> 
> It's not (yet), we have an open issue for that: [scylladb/scylla-docs#3787](https://github.com/scylladb/scylla-docs/issues/3787) . Some context should be available here: #8982

I meant the encoding. Since it uses IDL, that's fine.


> 
> > and why did we not use IDL?
> 
> The sync point description is serialized using IDL (not sure why I wrote "IMR" in the comment); the version field and the checksum is not. Not sure it would make sense to use IDL for the checksum as we need to parse it first before the rest.

Agree we can't use IDL (unless we make checksumming part of IDL). But then the wrapper around it should be documented.

What about compatibility? Can we take a sync point serialized in one node and pass it to another?

piodul commented:
> Agree we can't use IDL (unless we make checksumming part of IDL). But then the wrapper around it should be documented.

The comment at the top of `db/hints/sync_point.cc` serves as a description. Do you suggest to move it somewhere else, e.g. to `docs/dev/`?

> What about compatibility? Can we take a sync point serialized in one node and pass it to another?

Sync point encodes a set of positions in hint queues on some node. They are not meant to be passed from one node to another. If one tries to pass v2 sync point to another node then it will either fail due to unrecognized version, or due to host id mismatch (the host id is a part of the sync point).

The code in the PR is prepared to deserialize both V1 and V2. After upgrading a node, sync points from before restart will work. There might only be a problem with sync points created during downgrade, but... I'm not sure whether we should care much about this problem. The main purpose of sync points is to allow waiting until all current hints are replayed; in that case the user can generate a new sync point and wait for it again. We could avoid this with a cluster feature but I'm not sure whether it's worth introducing one.

kbr-scylla commented on db/hints/sync_point.cc:
```diff
@@ -69,14 +71,13 @@ static std::vector<sync_point::shard_rps> decode_one_type_v1(uint16_t shard_coun
-        throw std::runtime_error(format("Unsupported sync point format version: {}", int(raw[0])));
+    seastar::simple_memory_input_stream in{reinterpret_cast<const char*>(raw.data()), raw.size()};
+
+    uint8_t version = ser::serializer<uint8_t>::read(in);
```
The `raw.empty()` check is gone now. What will happen when it's empty and we do this `read`? If I understand correctly, we will get `std::out_of_range("deserialization buffer underflow")` (from `simple_memory_input_stream` implementation). I have to say I liked the previous error better -- "Could not decode the sync point - not a valid hex string".

patjed41 replied:
I've restored the `raw.empty()` check.

kbr-scylla commented on db/hints/sync_point.cc:
```diff
@@ -71,7 +71,8 @@ static std::vector<sync_point::shard_rps> decode_one_type_v1(uint16_t shard_coun
 sync_point sync_point::decode(sstring_view s) {
     bytes raw = base64_decode(s);
-    seastar::simple_memory_input_stream in{reinterpret_cast<const char*>(raw.data()), raw.size()};
+    sstring_view raw_s(reinterpret_cast<char*>(raw.data()), raw.size());
```
Why drop the `const`?

patjed41 replied:
Fixed.

kbr-scylla commented on db/hints/sync_point.cc:
```diff
@@ -135,12 +136,12 @@ sstring sync_point::encode() const {
     ser::serializer<sync_point_v1>::write(out, v1);
 
-    return base64_encode(serialized);
+    return base64_encode(bytes_view(reinterpret_cast<int8_t*>(serialized.data()), serialized.size()));
```
`reinterpret_cast<int8_t*>(serialized.data()), serialized.size()));`
-- doesn't this break type aliasing rules (i.e. results in undefined behavior)?

From https://en.cppreference.com/w/cpp/language/reinterpret_cast
> Any object pointer type T1* can be converted to another object pointer type cv T2*. (...) In any case, the resulting pointer may only be dereferenced safely if allowed by the type aliasing rules (see below). 

And the type aliasing rules:
> Whenever an attempt is made to read or modify the stored value of an object of type DynamicType through a glvalue of type AliasedType, the behavior is undefined unless one of the following is true:
> 
> - AliasedType and DynamicType are similar.
> - AliasedType is the (possibly cv-qualified) signed or unsigned variant of DynamicType.
> - AliasedType is std::byte, (since C++17) char, or unsigned char: this permits examination of the object representation of any object as an array of bytes. 
> 
> Informally, two types are similar if, ignoring top-level cv-qualification:
> 
> - they are the same type; or
> - they are both pointers, and the pointed-to types are similar; or
> - they are both pointers to member of the same class, and the types of the pointed-to members are similar; or 

So if I understand correctly, `DynamicType` is `char` (`serialized.data()` is `char*`) and `AliasedType` is `int8_t`. And this case is not covered by the type aliasing rules.

Maybe I'm misreading something though.

kbr-scylla replied:
BTW. I have no idea why `bytes` is defined as `basic_sstring<int8_t, ...>` and not `basic_sstring<char, ...>` (or better, `basic_sstring<std::byte, ...>`. Then we wouldn't have this problem.

piodul replied:
Hm, good catch. So it looks like the conversion `bytes -> sstring_view` is safe, but `sstring -> bytes_view` is not. In that case, I suggest to:

- Stick to `bytes` as the base storage,
- Create a `sstring_view` object based on the bytes (via reinterpret casting),
- Use either the bytes or the string view where appropriate.

patjed41 replied:
If I understand the discussion in #14689 correctly, the current conversion is safe, but I'll wait for confirmation.

kbr-scylla replied:
If I understand it correctly, the conversion would be safe if `sstring` was using `unsigned char` or `byte` in its definition, it doesn't work for `char`. But I'm not sure, this is a high level of sorcery.
cc @michoecho please advise

kbr-scylla replied:
In any case, I would feel much safer if we didn't have to depend on this rule.

And Il think that the refactor to use sstring instead of bytes is not worth it, it introduces a lot of "noise" to the PR just to save a single `reinterpret_cast`.

michoecho replied:
I confirm that this cast is technically illegal in the current revision of the C++ standard.
But also AFAIK it should not cause problems in any sane compiler.

kbr-scylla commented on db/hints/sync_point.cc:
```diff
@@ -17,13 +17,15 @@
+// Format V1/V2 (encoded in base64):
+//   uint8_t 0x01/0x02 - version of format
 //   sync_point_v1 - encoded using IMR
+//   uint64_t - checksum computed using the xxHash algorithm (present only in V2)
```
The comment would be more readable if "Format V2" had its own section IMO.

patjed41 replied:
Agree. I've also added a new comment:
```cpp
// Sync points can be encoded in two formats: V1 and V2. V2 extends V1 by adding
// a checksum. Currently, we use the V2 format, but sync points encoded in the V1
// format still can be safely decoded.
```
I think it's nice too understand this V1/V2 concept without having to analyze the code.

patjed41 commented:
v6:
- restored the `raw.empty()` check in `sync_point::decode`
- fixed the `const` dropping in `sstring_view raw_s(reinterpret_cast<const char*>(raw.data()), raw.size());`
- clarified comments about the V1 and V2 formats

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2483/
